File: src/main/java/org/apache/fineract/client/services/ProductsApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import rx.Observable;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.GetProductsTypeProductIdResponse;
import org.apache.fineract.client.models.GetProductsTypeResponse;
import org.apache.fineract.client.models.PostProductsTypeRequest;
import org.apache.fineract.client.models.PostProductsTypeResponse;
import org.apache.fineract.client.models.PutProductsTypeProductIdRequest;
import org.apache.fineract.client.models.PutProductsTypeProductIdResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface ProductsApi {
  /**
   * Create a Share Product
   * Creates a Share Product  Mandatory Fields: name, shortName, description, currencyCode, digitsAfterDecimal,inMultiplesOf, locale, totalShares, unitPrice, nominalShares,allowDividendCalculationForInactiveClients,accountingRule  Mandatory Fields for Cash based accounting (accountingRule &#x3D; 2): shareReferenceId, shareSuspenseId, shareEquityId, incomeFromFeeAccountId  Optional Fields: sharesIssued, minimumShares, maximumShares, minimumActivePeriodForDividends, minimumactiveperiodFrequencyType, lockinPeriodFrequency, lockinPeriodFrequencyType, marketPricePeriods, chargesSelected
   * @param type type (required)
   * @param postProductsTypeRequest  (required)
   * @return Observable&lt;PostProductsTypeResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @POST("products/{type}")
  Observable<PostProductsTypeResponse> createProduct(
    @retrofit2.http.Path("type") String type, @retrofit2.http.Body PostProductsTypeRequest postProductsTypeRequest
  );

  /**
   * 
   * 
   * @param type type (required)
   * @param productId productId (required)
   * @param command command (optional)
   * @return Observable&lt;String&gt;
   */
  @POST("products/{type}/{productId}")
  Observable<String> handleCommands3(
    @retrofit2.http.Path("type") String type, @retrofit2.http.Path("productId") Long productId, @retrofit2.http.Query("command") String command
  );

  /**
   * List Share Products
   * Lists Share Products  Mandatory Fields: limit, offset  Example Requests:  shareproducts
   * @param type type (required)
   * @param offset offset (optional)
   * @param limit limit (optional)
   * @return Observable&lt;GetProductsTypeResponse&gt;
   */
  @GET("products/{type}")
  Observable<GetProductsTypeResponse> retrieveAllProducts(
    @retrofit2.http.Path("type") String type, @retrofit2.http.Query("offset") Integer offset, @retrofit2.http.Query("limit") Integer limit
  );

  /**
   * Retrieve a Share Product
   * Retrieves a Share Product  Example Requests:  products/share/1   products/share/1?template&#x3D;true
   * @param productId productId (required)
   * @param type type (required)
   * @return Observable&lt;GetProductsTypeProductIdResponse&gt;
   */
  @GET("products/{type}/{productId}")
  Observable<GetProductsTypeProductIdResponse> retrieveProduct(
    @retrofit2.http.Path("productId") Long productId, @retrofit2.http.Path("type") String type
  );

  /**
   * 
   * 
   * @param type type (required)
   * @return Observable&lt;String&gt;
   */
  @GET("products/{type}/template")
  Observable<String> retrieveTemplate12(
    @retrofit2.http.Path("type") String type
  );

  /**
   * Update a Share Product
   * Updates a Share Product
   * @param type type (required)
   * @param productId productId (required)
   * @param putProductsTypeProductIdRequest  (required)
   * @return Observable&lt;PutProductsTypeProductIdResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @PUT("products/{type}/{productId}")
  Observable<PutProductsTypeProductIdResponse> updateProduct(
    @retrofit2.http.Path("type") String type, @retrofit2.http.Path("productId") Long productId, @retrofit2.http.Body PutProductsTypeProductIdRequest putProductsTypeProductIdRequest
  );

}

File: src/main/kotlin/org/openapitools/client/apis/ProductsApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.GetProductsTypeProductIdResponse
import org.openapitools.client.models.GetProductsTypeResponse
import org.openapitools.client.models.PostProductsTypeRequest
import org.openapitools.client.models.PostProductsTypeResponse
import org.openapitools.client.models.PutProductsTypeProductIdRequest
import org.openapitools.client.models.PutProductsTypeProductIdResponse
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface ProductsApi {
    /**
     * Create a Share Product
     * Creates a Share Product  Mandatory Fields: name, shortName, description, currencyCode, digitsAfterDecimal,inMultiplesOf, locale, totalShares, unitPrice, nominalShares,allowDividendCalculationForInactiveClients,accountingRule  Mandatory Fields for Cash based accounting (accountingRule &#x3D; 2): shareReferenceId, shareSuspenseId, shareEquityId, incomeFromFeeAccountId  Optional Fields: sharesIssued, minimumShares, maximumShares, minimumActivePeriodForDividends, minimumactiveperiodFrequencyType, lockinPeriodFrequency, lockinPeriodFrequencyType, marketPricePeriods, chargesSelected
     * Responses:
     *  - 200: OK
     *
     * @param type type
     * @param postProductsTypeRequest
     * @return [PostProductsTypeResponse]
     */
    @POST("v1/products/{type}")
    suspend fun createProduct(
        @Path("type") type: String,
        @Body postProductsTypeRequest: PostProductsTypeRequest
    ): PostProductsTypeResponse

    /**
     *
     *
     * Responses:
     *  - 0: default response
     *
     * @param type type
     * @param productId productId
     * @param command command (optional)
     * @return [kotlin.String]
     */
    @POST("v1/products/{type}/{productId}")
    suspend fun handleCommands3(
        @Path("type") type: String,
        @Path("productId") productId: Long,
        @Query("command") command: String? = null
    ): String

    /**
     * List Share Products
     * Lists Share Products  Mandatory Fields: limit, offset  Example Requests:  shareproducts
     * Responses:
     *  - 200: OK
     *
     * @param type type
     * @param offset offset (optional)
     * @param limit limit (optional)
     * @return [GetProductsTypeResponse]
     */
    @GET("v1/products/{type}")
    suspend fun retrieveAllProducts(
        @Path("type") type: String,
        @Query("offset") offset: Int? = null,
        @Query("limit") limit: Int? = null
    ): GetProductsTypeResponse

    /**
     * Retrieve a Share Product
     * Retrieves a Share Product  Example Requests:  products/share/1   products/share/1?template&#x3D;true
     * Responses:
     *  - 200: OK
     *
     * @param productId productId
     * @param type type
     * @return [GetProductsTypeProductIdResponse]
     */
    @GET("v1/products/{type}/{productId}")
    suspend fun retrieveProduct(
        @Path("productId") productId: Long,
        @Path("type") type: String
    ): GetProductsTypeProductIdResponse

    /**
     *
     *
     * Responses:
     *  - 0: default response
     *
     * @param type type
     * @return [kotlin.String]
     */
    @GET("v1/products/{type}/template")
    suspend fun retrieveTemplate13(@Path("type") type: String): String

    /**
     * Update a Share Product
     * Updates a Share Product
     * Responses:
     *  - 200: OK
     *
     * @param type type
     * @param productId productId
     * @param putProductsTypeProductIdRequest
     * @return [PutProductsTypeProductIdResponse]
     */
    @PUT("v1/products/{type}/{productId}")
    suspend fun updateProduct(
        @Path("type") type: String,
        @Path("productId") productId: Long,
        @Body putProductsTypeProductIdRequest: PutProductsTypeProductIdRequest
    ): PutProductsTypeProductIdResponse

}
